Use conditional mutations for AccumuloStore. Add checks for status and putRepo

DomGarguilo opened the pull request:
Closes #4114

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/FateMutatorImpl.java:
```diff
@@ -132,12 +142,29 @@ public FateMutator<T> delete() {
     return this;
   }
 
+  /**
```
The status column is expected to exist for the lifetime of a fate operation.  As a start of using conditional mutations we could build something around and always checking the status column when an update is made.  This will check that row is in a minimally acceptable state for making an update.  We could add more conditions later.

Could add a method like the following to FateMutatorImpl

```java
/**
 * Requires the status column to be set to one of the passed in statuses in order for an update to be made.  If no statuses are passed in, then the status column must be absent in order for an update to be made.
 */
 public FateMutator<T> requireStatus(TStatus ... statuses) {}
```

With a method like the above could do the following when adding repo.

```java
      FateMutator<T> fateMutator = newMutator(tid).requireStatus(SUBMITTED, IN_PROGRESS);
      fateMutator.putRepo(top.map(t -> t + 1).orElse(1), repo).mutate();
```

When adding a completely new fate tx could do the following which would require the status column to be absent.

```java
  @Override
  public long create() {
    long tid = RANDOM.get().nextLong() & 0x7fffffffffffffffL;

    newMutator(tid).requireStatus().putStatus(TStatus.NEW).putCreateTime(System.currentTimeMillis()).mutate();

    return tid;
  }
```

DomGarguilo replied:
If I am understanding things correctly, this new method would check the value of the status column. Is it possible to have a set of acceptable values for a Condition? To me it seems like you can only set one value with `Condition.setValue()`.

keith-turner replied:
To to check for a set of possible values can do the following

 * Set an iterator on the condition that maps the value to something like  `present` or `not present` depending on if the value is in the set.
 * Set the expected value on the condition to `present`

DomGarguilo replied:
Okay I added this in e5fa0da. I think there is an issue though but I can't figure out whats causing it. See my next comment please.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/FateMutatorImpl.java:
```diff
@@ -105,13 +108,20 @@ public FateMutator<T> putTxInfo(TxInfo txInfo, byte[] data) {
-    mutation.put(RepoColumnFamily.NAME, invertRepo(position), new Value(serialize(repo)));
+    final Text cq = invertRepo(position);
+    // ensure this repo is not already set
+    mutation.addCondition(new Condition(RepoColumnFamily.NAME, cq));
```
We could add a stronger check that ensure the current max repo in storage is `position - 1`.  This could be done with a specialized iterator.  Or could have to conditions like :

 * A condition that checks position is absent
 * A condition that checks that position -1 is present.

DomGarguilo replied:
Using conditions, is there a way to verify that a column is present without knowing its value? To me it looks like we can only verify that a columns exists and verify that a column has a specific value, not that a column exists with any value.

DomGarguilo replied:
Made a note about this in the ticket: https://github.com/apache/accumulo/issues/4114#issuecomment-1900971079. Going to try to merge in what I have as is and handle that in another PR.

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateMutatorImplIT.java:
```diff
@@ -88,6 +89,65 @@ public void putRepo() throws Exception {
+
+      final long tid = RANDOM.get().nextLong() & 0x7fffffffffffffffL;
+
+      // use require status passing all statuses. without the status column present this should fail
+      // FateMutatorImpl<FateIT.TestEnv> fateMutator = new FateMutatorImpl<>(context, table, tid);
+      // assertThrows(IllegalStateException.class,
+      // () -> fateMutator.requireStatus(ReadOnlyFateStore.TStatus.values())
+      // .putStatus(ReadOnlyFateStore.TStatus.NEW).mutate());
+      //
+      // logAllEntriesInTable(table, client); // this prints nothing on the table as we expect
+
+      // use require status without passing any statuses to require that the status column is absent
+      FateMutatorImpl<FateIT.TestEnv> fateMutator0 = new FateMutatorImpl<>(context, table, tid);
+      fateMutator0.requireStatus().putStatus(ReadOnlyFateStore.TStatus.NEW).mutate();
```
Right now with this portion commented out, the test passes. The bottom portion adds a `requireStatus` with no params which should ensure that the status column does not exist. The weird part is, when i uncomment the top portion, it causes this check to fail even though (as verified by printing all entries on the table) the status column still does not exist. I am not sure what is causing the condition to fail at this point.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/StatusMappingIterator.java:
```diff
@@ -0,0 +1,134 @@
+      is.addOption(STATUS_SET_KEY, encodeStatuses(statuses));
+
+      // The iterator will map the status to "present" if it's in the acceptable set
+      return CONDITION.setValue(PRESENT).setIterators(is);
```
This is modifying the static variable CONDITION, so it will impact other uses of the variable.  The `setValue()` method does not create a new instance of Condition.

DomGarguilo replied:
Whoops, that was the issue. Thanks.

DomGarguilo commented:
Marked this PR as ready for review. Leaving other checks to be done in follow on PRs.

In this PR:
* Converted FateMutatorImpl to use ConditionalMutations
* Added a method (`FateMutatorImpl.requireStatus`) add a condition that checks the status value is what we expect and use it in `AccumuloStore.create()`. Created a new Iterator to make this possible
* Added a condition to `putRepo()` to make sure the spot we are inserting into is empty.
* Added a new IT to test the two mentioned added checks.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/FateMutatorImpl.java:
```diff
@@ -132,12 +140,30 @@ public FateMutator<T> delete() {
-      writer.addMutation(mutation);
-    } catch (Exception e) {
-      throw new IllegalStateException(e);
+    try (ConditionalWriter writer = context.createConditionalWriter(tableName)) {
```
For this method if there are not conditions, we could use a batch writer for the mutation instead of adding the condition for them empty column.

```suggestion
    if(mutation.getConditions().isEmpty()) {
       // use batch writer
    } else {
       // use conditional writer
    }
```

DomGarguilo replied:
Fixed in d8a3a05

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/AccumuloStore.java:
```diff
@@ -70,7 +70,8 @@ public AccumuloStore(ClientContext context) {
     long tid = RANDOM.get().nextLong() & 0x7fffffffffffffffL;
 
-    newMutator(tid).putStatus(TStatus.NEW).putCreateTime(System.currentTimeMillis()).mutate();
+    newMutator(tid).requireStatus().putStatus(TStatus.NEW).putCreateTime(System.currentTimeMillis())
```
We need to retry when this fails because the conditiona was not accepted.  Not sure of the best way to do this.  Could have a special exception and catch that OR make mutate() return something.  Also need to test this if possible.

Maybe we could make the next id a function that can be overridden by a test class.

```java
  // exist to be overridden by test so can test trying to create existing id
 protected long getNextId() {
    return RANDOM.get().nextLong() & 0x7fffffffffffffffL;
}

public long create() {
   long tid = getNextId();

   while(true) {
      // attempt to persist id
      // if successful return tid
     /// if not successful then try again  : tid = getNextId(); 
      //sleep or use retry obj
    }

```

Feel free to ignore the comments about testing. Not sure about my testing comment, maybe testing needs to be at a lower level like is already being done. Even w/o the changes for testing, do need to retry somehow.

DomGarguilo replied:
This is a good idea. I was also thinking about something similar: how we can't tell which condition attached to a mutation has caused it to fail. It would be nice to have something like an optional message that could be added to the ConditionalWriter.Result object so that when the result is returned, we could know what caused the mutation to fail.

As things are now though, I am not sure how we could trigger a retry, since there could be several things that cause an exception to be thrown, we don't know if its this case where we want a retry to happen.

keith-turner replied:
I experimented with making mutate() return something, but that seemed a bit dangerous as it would be easy to ignore the return value.  Realized maybe adding a new `tryMutate()` method might be good because the prefix `try` implies a return value, this is following some of the methods in the java api that have a `try` prefix.  Below is the commit where I experimented with this.

https://github.com/keith-turner/accumulo/commit/f70f6bc65a879515bc803358d5219cafbb90ec57

It also adds a new enum because we only want to handle a subset of the statuses that conditional writer can return.

keith-turner replied:
Also I was curious what the potential for a collision was and found : 

https://en.wikipedia.org/wiki/Birthday_attack

which has a nice table.  If I am reading the table correctly, for a 64 bit integer if there are 6.1M entries in the fate table then we would have a 1 in 1M chance of a collision.

DomGarguilo replied:
Added this in 9240cad

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/AccumuloStoreIT.java:
```diff
@@ -0,0 +1,94 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AccumuloStoreIT extends SharedMiniClusterBase {
```
This is a nice test.  If possible would be nice to eventually run the test against Zookeeper also like #4202.  If it makes sense could be a follow on.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/StatusMappingIterator.java:
```diff
@@ -0,0 +1,148 @@
+      String currentValue = source.getTopValue().toString();
+      mappedValue =
+          acceptableStatuses.contains(currentValue) ? new Value(PRESENT) : new Value(ABSENT);
+    }
```
```suggestion
    } else{
      mappedValue = null;
    }
```

DomGarguilo replied:
49f6f69

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/StatusMappingIterator.java:
```diff
@@ -0,0 +1,148 @@
+
+  @Override
+  public Value getTopValue() {
+    return mappedValue;
```
This makes things fail fast when mappedValue is not set

```suggestion
    return Objects.requireNonnull(mappedValue);
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateMutatorImplIT.java:
```diff
@@ -0,0 +1,154 @@
+      // use require status without passing any statuses to require that the status column is absent
+      FateMutatorImpl<FateIT.TestEnv> fateMutator0 = new FateMutatorImpl<>(context, table, tid);
+      fateMutator0.requireStatus().putStatus(ReadOnlyFateStore.TStatus.NEW).mutate();
+
```
would be good to read the status after each change or attempted change to ensure its as expected

DomGarguilo replied:
49f6f69
